FOLIO: smarter handling of request types in placeHold.

demiankatz opened the pull request:
This PR makes two improvements to FOLIO's placeHold functionality:

- It fixes a bug where default request type was set inappropriately for title level holds (because title-level hold data does not include availability status)
- It adds a mechanism to try multiple request types, in case one type fails due to rules

TODO
- [x] Add test coverage of new functionality before merging.

demiankatz committed:
FOLIO: smarter handling of request types.

demiankatz committed:
Delete module/VuFind/tests/.phpunit.cache/test-results

demiankatz committed:
Satisfy phpstan.

maccabeelevine commented on config/vufind/Folio.ini:
```diff
@@ -116,10 +116,19 @@ defaultRequiredDate = 0:1:0
+; "Recall"
 ;default_request = Hold
 
+; It is possible that the default_request method will not work in every situation; if
```
"will not work in every situation" is true but implies (IMHO) a problem. rather than something more likely intentional.  Maybe add something like "such as due to specific circulation rules."

demiankatz replied:
Thanks, @maccabeelevine. Comment improved in commit e96af67887f4ff40d02ec7960ceb8f8475a937fc.

maccabeelevine reviewed (approved):
The logic looks fine.  I don't have an environment available with title-level holds to test in unfortunately.

demiankatz committed:
Improve comment.

demiankatz committed:
Merge branch 'dev' into smarter-folio-requests

demiankatz committed:
Add test; fix error handling.

demiankatz commented:
@maccabeelevine, if you don't mind taking one more quick look here: I discovered while writing tests that the hold response validation code was a bit wonky. I refactored it as part of commit da11ca1c7932bd642ed7a1ef14bfb47758b97fe2. It would be good to get a double-check on that work before merging this.

To clarify what has changed: previously, we assumed we had parseable JSON as long as we had a good HTTP response code, and we assumed that the JSON response included a status. If these assumptions were not met, we'd access an undefined property and fail. Now we check first, and fail if expected things are missing.

Follow-up: the problem I addressed was actually triggered by a problem in the test framework, addressed by #3382... but I think it's still sensible to adjust the processing to be more robust, even if the issue was not triggered by real-world causes.

demiankatz committed:
Fix fixture.

maccabeelevine reviewed (approved):
> To clarify what has changed: previously, we assumed we had parseable JSON as long as we had a good HTTP response code, and we assumed that the JSON response included a status. If these assumptions were not met, we'd access an undefined property and fail. Now we check first, and fail if expected things are missing.

The change makes sense to me.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into smarter-folio-requests
